Public/Src/Cache/ContentStore/Distributed/Redis/RedisContentLocationStore.cs

```diff
@@ -1306,6 +1306,24 @@ public Task<Result<byte[]>> GetBlobAsync(OperationContext context, ContentHash h
             return _blobAdapter.GetBlobAsync(context, hash);
         }
 
+        /// <inheritdoc />
+        public MachineLocation GetRandomMachineLocation(MachineLocation except)
+        {
+            if (_idsByLocation.Keys.Any(location => !location.Equals(except)))
+            {
+                MachineLocation location;
+                do
+                {
+                    location = _idsByLocation.Keys.ElementAt(ThreadSafeRandom.Generator.Next(_idsByLocation.Keys.Count));
+                }
+                while (location.Equals(except));
+
+                return location;
+            }

```

Comment:
If we have two similar implementations we can add a helper method that will give us machine location from a collection of machine locations to avoid code duplication.

Reply:
They ended up being slightly different because in one case we get an array which can contain the 'default' value, and we need to make sure that we don't return that.

Reply:
They ended up being slightly different because in one case we get an array which can contain the 'default' value, and we need to make sure that we don't return that.

---
In reply to: [312219858](https://github.com/microsoft/BuildXL/pull/705#discussion_r312219858) [](ancestors = 312219858)

Reply:
Reactivating. I think we can unify these. Let's chat.

---
In reply to: [312661031](https://github.com/microsoft/BuildXL/pull/705#discussion_r312661031) [](ancestors = 312661031,312219858)